sql/src/main/java/io/crate/execution/dml/upsert/TransportShardInsertAction.java

```diff
@@ -0,0 +1,205 @@
+        final ReturnValueGen returnValueGen;
+
+        final ShardInsertRequest.DuplicateKeyAction duplicateKeyAction;
+
+        public Context(InsertSourceGen insertSourceGen, ReturnValueGen returnValueGen, ShardInsertRequest.DuplicateKeyAction duplicateKeyAction) {
+            this.insertSourceGen = insertSourceGen;
+            this.returnValueGen = returnValueGen;
+            this.duplicateKeyAction = duplicateKeyAction;
+        }
+    }
+
+    @Inject
+    public TransportShardInsertAction(ThreadPool threadPool,
+                                      ClusterService clusterService,
+                                      TransportService transportService,
+                                      SchemaUpdateClient schemaUpdateClient,
+                                      TasksService tasksService,
+                                      IndicesService indicesService,
+                                      ShardStateAction shardStateAction,
+                                      Functions functions,
+                                      Schemas schemas,
+                                      IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(
+            ACTION_NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            schemaUpdateClient,
+            tasksService,
+            indicesService,
+            shardStateAction,
+            ShardInsertRequest::new,
+            indexNameExpressionResolver
+        );
+        this.schemas = schemas;
+        this.functions = functions;
+        tasksService.addListener(this);
+    }
+
+    public Context buildContext(ShardInsertRequest request) {
+        String indexName = request.index();
+        DocTableInfo tableInfo = schemas.getTableInfo(RelationName.fromIndexName(indexName), Operation.INSERT);
+        Reference[] insertColumns = request.insertColumns();
+        GeneratedColumns.Validation valueValidation = request.validateConstraints()
+            ? GeneratedColumns.Validation.VALUE_MATCH
+            : GeneratedColumns.Validation.NONE;
+
+        TransactionContext txnCtx = TransactionContext.of(request.sessionSettings());
+        InsertSourceGen insertSourceGen = insertColumns == null
+            ? null
+            : InsertSourceGen.of(txnCtx, functions, tableInfo, indexName, valueValidation, Arrays.asList(insertColumns));
+
+        ReturnValueGen returnValueGen = request.returnValues() == null
+            ? null
+            : new ReturnValueGen(functions, txnCtx, tableInfo, request.returnValues());
+
+        return new Context(insertSourceGen, returnValueGen, request.duplicateKeyAction());
+    }
+
+
+    @Override
+    IndexItemResponse processItem(ShardInsertRequest.Item item, Context context, IndexShard indexShard) throws Exception {
+        VersionConflictEngineException lastException = null;
+        boolean isRetry;
+        for (int retryCount = 0; retryCount < MAX_RETRY_LIMIT; retryCount++) {
```

Comment:
The main reason why this loop is still here and not in `TransportShardIndexAction` is because i need to have it for failure handling in some cases.